core/sequencer/server.go

```diff
@@ -108,39 +109,85 @@ func NewServer(
 
 // RunBatch reads mutations out of the queue and calls CreateEpoch.
 func (s *Server) RunBatch(ctx context.Context, in *spb.RunBatchRequest) (*empty.Empty, error) {
-	batch, err := s.queue.ReadQueue(ctx, in.DomainId, in.MaxBatch)
+	// Get the previous and current high water marks.
+	domain, err := s.ktServer.GetDomain(ctx, &ktpb.GetDomainRequest{DomainId: in.DomainId})
 	if err != nil {
-		return nil, status.Errorf(codes.Internal, "ReadQueue(): %v", err)
+		return nil, err
+	}
+	mapClient, err := tclient.NewMapClientFromTree(s.tmap, domain.Map)
+	if err != nil {
+		return nil, err
+	}
+	latestMapRoot, err := mapClient.GetAndVerifyLatestMapRoot(ctx)
+	if err != nil {
+		return nil, err
+	}
+	var lastMeta spb.MapMetadata
+	if err := proto.Unmarshal(latestMapRoot.Metadata, &lastMeta); err != nil {
+		return nil, err
+	}
+	high, err := s.queue.HighWatermark(ctx, in.DomainId)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "HighWatermark(): %v", err)
+	}
+
+	// TODO(gbelvin): If time since last map revision > max timeout, run batch.
+	// TODO(#1047): If time since oldest queue item > max latency has elapsed, run batch.
+	// TODO(gbelvin): If count items > max_batch, run batch.
+
+	// Count items to be processed.  Unfortunately, this means we will be
+	// reading the items to be processed twice.  Once, here in RunBatch and
+	// again in CreateEpoch.
+	metadata := &spb.MapMetadata{
+		Source: &spb.MapMetadata_SourceSlice{
+			LowestWatermark:  lastMeta.GetSource().GetHighestWatermark(),
+			HighestWatermark: high,
+		},
 	}
-	if int32(len(batch)) < in.MinBatch {
-		// TODO(gbelvin): Process remaining messages after a timeout.
```

Comment:
This `TODO(#1047)` is probably worth keeping (in addition to the one in the sequencer_api.proto file)

Reply:
The `TODO` is still there, just a few lines up

Reply:
Ah, sorry, deficiencies of the GitHub UI...